[SPARK-17495] [SQL] Support date, timestamp and interval types in Hive hash

tejasapatil opened the pull request:
## What changes were proposed in this pull request?

- Timestamp hashing is done as per [TimestampWritable.hashCode()](https://github.com/apache/hive/blob/ff67cdda1c538dc65087878eeba3e165cf3230f4/serde/src/java/org/apache/hadoop/hive/serde2/io/TimestampWritable.java#L406) in Hive
- Interval hashing is done as per [HiveIntervalDayTime.hashCode()](https://github.com/apache/hive/blob/ff67cdda1c538dc65087878eeba3e165cf3230f4/storage-api/src/java/org/apache/hadoop/hive/common/type/HiveIntervalDayTime.java#L178). Note that there are inherent differences in how Hive and Spark store intervals under the hood which limits the ability to be in completely sync with hive's hashing function. I have explained this in the method doc.
- Date type was already supported. This PR adds test for that.

## How was this patch tested?

Added unit tests

tejasapatil committed:
[SPARK-17495] [SQL] Support date, timestamp datatypes in Hive hash

tejasapatil committed:
minor refac

tejasapatil commented:
ok to test

SparkQA commented:
**[Test build #73459 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73459/testReport)** for PR 17062 at commit [`332475c`](https://github.com/apache/spark/commit/332475c1641f61080aa41dda9f1ceec237351d75).

SparkQA commented:
**[Test build #73459 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73459/testReport)** for PR 17062 at commit [`332475c`](https://github.com/apache/spark/commit/332475c1641f61080aa41dda9f1ceec237351d75).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73459/
Test FAILed.

tejasapatil committed:
Cast long to int for handling

Caused by: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 84, Column 12: Assignment conversion not possible from type "long" to type "int"
    at org.codehaus.janino.UnitCompiler.compileError(UnitCompiler.java:11004)
    at org.codehaus.janino.UnitCompiler.assignmentConversion(UnitCompiler.java:9885)
    at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:3187)

/* 084 */     childHash = org.apache.spark.sql.catalyst.expressions.HiveHashFunction.hashTimestamp(-8951713928982000L);value = (31 * value) + childHash;

tejasapatil commented:
Jenkins retest this please

SparkQA commented:
**[Test build #73472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73472/testReport)** for PR 17062 at commit [`e050a50`](https://github.com/apache/spark/commit/e050a5000d0bb4620d9eb7890da5206e4e36ce09).

SparkQA commented:
**[Test build #73472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73472/testReport)** for PR 17062 at commit [`e050a50`](https://github.com/apache/spark/commit/e050a5000d0bb4620d9eb7890da5206e4e36ce09).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73472/
Test PASSed.

tejasapatil commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:
```diff
@@ -169,6 +171,96 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+      checkHiveHash(CalendarInterval.fromString(interval), CalendarIntervalType, expected)
+    }
+
+    checkHiveHashForTimestampType("interval 1 day", 3220073)
```
SELECT HASH ( INTERVAL '1' DAY );

tejasapatil commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:
```diff
@@ -169,6 +171,96 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+    }
+
+    checkHiveHashForTimestampType("interval 1 day", 3220073)
+    checkHiveHashForTimestampType("interval 6 day 15 hour", 21202073)
```
SELECT HASH ( INTERVAL '1' DAY + INTERVAL '15' HOUR );

tejasapatil commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:
```diff
@@ -169,6 +171,96 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+
+    checkHiveHashForTimestampType("interval 1 day", 3220073)
+    checkHiveHashForTimestampType("interval 6 day 15 hour", 21202073)
+    checkHiveHashForTimestampType("interval -23 day 56 hour -1111113 minute 9898989 second",
```
SELECT HASH ( INTERVAL '-23' DAY + INTERVAL '56' HOUR + INTERVAL '-1111113' MINUTE + INTERVAL '9898989' SECOND );

tejasapatil commented:
@gatorsmile : can you please review this PR ?

tejasapatil commented:
Updated comments with the corresponding hive queries used to generate the expected outputs.

gatorsmile commented:
Will review it tonight. Thanks!

gatorsmile commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:
```diff
@@ -169,6 +171,96 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+    checkHiveHashForTimestampType("interval 1 day", 3220073)
+    checkHiveHashForTimestampType("interval 6 day 15 hour", 21202073)
+    checkHiveHashForTimestampType("interval -23 day 56 hour -1111113 minute 9898989 second",
+      -2128468593)
```
Coud you add more test cases?

```
    checkHiveHashForTimestampType("interval 0 day 0 hour 0 minute 0 second", 23273)
    checkHiveHashForTimestampType("interval 0 day 0 hour", 23273)
    checkHiveHashForTimestampType("interval -1 day", 3220036)
```

tejasapatil replied:
added

gatorsmile commented:
It sounds like no test case covers nanosecond for INTERVAL

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:
```diff
@@ -732,6 +741,38 @@ object HiveHashFunction extends InterpretedHashFunction {
+   * the interval and is unified.
+   */
+  def hashCalendarInterval(calendarInterval: CalendarInterval): Long = {
+    val totalSeconds = calendarInterval.milliseconds() / 1000
```
How does Hive deal with nanoseconds, if we divide it by 1000?

tejasapatil replied:
Spark's CalendarInterval has precision upto microseconds while Hive can have precision upto nanoseconds. So, there is no way for us to support that in the hashing function. I have documented this in the PR.

tejasapatil committed:
review comment #1

tejasapatil reviewed (commented):
Updated the PR. Added more tests.

SparkQA commented:
**[Test build #74278 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74278/testReport)** for PR 17062 at commit [`fd0330d`](https://github.com/apache/spark/commit/fd0330d09551b69770a9fc586145b5547d3d44dc).

tejasapatil commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:
```diff
@@ -168,6 +170,208 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+    intercept[TestFailedException](checkHiveHashForTimestampType("2017-02-24 10:56:29.11111111", 0))
+  }
+
+  test("hive-hash for CalendarInterval type") {
```
Hive queries for all the tests below. Outputs are generated by running against Hive-1.2.1

```
// ----- MICROSEC -----
SELECT HASH(interval_day_time("0 0:0:0.000001") );
SELECT HASH(interval_day_time("-0 0:0:0.000001") );
SELECT HASH(interval_day_time("0 0:0:0.000000") );
SELECT HASH(interval_day_time("0 0:0:0.000999") );
SELECT HASH(interval_day_time("-0 0:0:0.000999") );

// ----- MILLISEC -----
SELECT HASH(interval_day_time("0 0:0:0.001") );
SELECT HASH(interval_day_time("-0 0:0:0.001") );
SELECT HASH(interval_day_time("0 0:0:0.000") );
SELECT HASH(interval_day_time("0 0:0:0.999") );
SELECT HASH(interval_day_time("-0 0:0:0.999") );

// ----- SECOND -----
SELECT HASH( INTERVAL '1' SECOND);
SELECT HASH( INTERVAL '-1' SECOND);
SELECT HASH( INTERVAL '0' SECOND);
SELECT HASH( INTERVAL '2147483647' SECOND);
SELECT HASH( INTERVAL '-2147483648' SECOND);

// ----- MINUTE -----
SELECT HASH( INTERVAL '1' MINUTE);
SELECT HASH( INTERVAL '-1' MINUTE);
SELECT HASH( INTERVAL '0' MINUTE);
SELECT HASH( INTERVAL '2147483647' MINUTE);
SELECT HASH( INTERVAL '-2147483648' MINUTE);

// ----- HOUR -----
SELECT HASH( INTERVAL '1' HOUR);
SELECT HASH( INTERVAL '-1' HOUR);
SELECT HASH( INTERVAL '0' HOUR);
SELECT HASH( INTERVAL '2147483647' HOUR);
SELECT HASH( INTERVAL '-2147483648' HOUR);

// ----- DAY -----
SELECT HASH( INTERVAL '1' DAY);
SELECT HASH( INTERVAL '-1' DAY);
SELECT HASH( INTERVAL '0' DAY);
SELECT HASH( INTERVAL '106751991' DAY);
SELECT HASH( INTERVAL '-106751991' DAY);

// ----- MIX -----
SELECT HASH( INTERVAL '0' DAY );
SELECT HASH( INTERVAL '0' DAY + INTERVAL '0' HOUR );
SELECT HASH( INTERVAL '0' DAY + INTERVAL '0' HOUR + INTERVAL '0' MINUTE);
SELECT HASH( INTERVAL '0' DAY + INTERVAL '0' HOUR + INTERVAL '0' MINUTE + INTERVAL '0' SECOND);
SELECT HASH(interval_day_time("0 0:0:0.000") );
SELECT HASH(interval_day_time("0 0:0:0.000000") );

SELECT HASH( INTERVAL '6' DAY + INTERVAL '15' HOUR );
SELECT HASH( INTERVAL '5' DAY + INTERVAL '4' HOUR + INTERVAL '8' MINUTE);
SELECT HASH ( INTERVAL '-23' DAY + INTERVAL '56' HOUR + INTERVAL '-1111113' MINUTE + INTERVAL '9898989' SECOND );
SELECT HASH(interval_day_time("66 12:39:23.987") );
SELECT HASH(interval_day_time("66 12:39:23.987123") );
```

SparkQA commented:
**[Test build #74278 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74278/testReport)** for PR 17062 at commit [`fd0330d`](https://github.com/apache/spark/commit/fd0330d09551b69770a9fc586145b5547d3d44dc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74278/
Test PASSed.

tejasapatil commented:
cc @gatorsmile

gatorsmile commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:
```diff
@@ -168,6 +170,208 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+
+    // Out of range for both Hive and Spark
+    // Hive throws an exception. Spark overflows and returns wrong output
+    // checkHiveHashForIntervalType("interval 9999999999 day", -4767228)
```
Should we fix it before this PR?

tejasapatil replied:
In case of Spark SQL, the query with fails with exception (see below). However for the test case since I am by-passing and creating raw interval object which does not go through that check

```
scala> hc.sql("SELECT interval 9999999999 day ").show
org.apache.spark.sql.catalyst.parser.ParseException:
Error parsing interval string: day 9999999999 outside range [-106751991, 106751991](line 1, pos 16)

== SQL ==
SELECT interval 9999999999 day
```

```
scala> df.select("INTERVAL 9999999999 day").show()
org.apache.spark.sql.AnalysisException: cannot resolve '`INTERVAL 9999999999 day`' given input columns: [key, value];;
'Project ['INTERVAL 9999999999 day]
```

gatorsmile commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:
```diff
@@ -168,6 +170,208 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+
+    // Out of range for both Hive and Spark
+    // Hive throws an exception. Spark overflows and returns wrong output
+    // checkHiveHashForIntervalType("interval 9999999999 day", -4767228)
```
The unit sounds incorrect. The same to the other case

tejasapatil replied:
fixed

tejasapatil committed:
review comment #2

SparkQA commented:
**[Test build #74414 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74414/testReport)** for PR 17062 at commit [`332686a`](https://github.com/apache/spark/commit/332686acd902c2b05bf48b848ece0860de172355).

gatorsmile commented:
I did the same check. The results of Hive 2.0 exactly match the hard-coded values.

gatorsmile commented:
@tejasapatil Thanks for your work! Could you add a comment in the `hash` function? The caller of `hash` needs to check the validity of input values. 

LGTM pending test.

tejasapatil committed:
review #3

tejasapatil commented:
@gatorsmile : Thanks for the review :) Added method doc for `hash()` with the comment as suggested.

SparkQA commented:
**[Test build #74416 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74416/testReport)** for PR 17062 at commit [`8a5f200`](https://github.com/apache/spark/commit/8a5f200427d616913ea506e34a11a5a698a20c8a).

SparkQA commented:
**[Test build #74414 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74414/testReport)** for PR 17062 at commit [`332686a`](https://github.com/apache/spark/commit/332686acd902c2b05bf48b848ece0860de172355).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74414/
Test PASSed.

SparkQA commented:
**[Test build #74416 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74416/testReport)** for PR 17062 at commit [`8a5f200`](https://github.com/apache/spark/commit/8a5f200427d616913ea506e34a11a5a698a20c8a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74416/
Test PASSed.

gatorsmile commented:
Thanks! Merging to master.
